File: cpp/endpoints/control_selection/set_mode.cpp
#include "rclcpp/rclcpp.hpp"
#include "clearpath_control_msgs/srv/set_control_mode.hpp"
#include "clearpath_control_msgs/msg/control_mode.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <yaml-cpp/yaml.h>

// enter your control mode
const int CONTROL_MODE = clearpath_control_msgs::msg::ControlMode::MANUAL; // options: MANUAL, AUTONOMOUS, NEUTRAL. See clearpath_control_msgs/msg/ControlMode.msg for more details.

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
  std::string namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("set_control_mode_client");
  rclcpp::Client<clearpath_control_msgs::srv::SetControlMode>::SharedPtr client =
    node->create_client<clearpath_control_msgs::srv::SetControlMode>("/" + namespace_ + "/" + std::string("control_selection/set_mode"));

  auto request = std::make_shared<clearpath_control_msgs::srv::SetControlMode::Request>();
  request->mode.mode = CONTROL_MODE;

  while (!client->wait_for_service(2s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("set_control_mode_client"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("set_control_mode_client"), "/%s/control_selection/set_mode service not available, waiting again...", namespace_.c_str());
  }

  auto result = client->async_send_request(request);

  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    if (result.get()->success)
    {
      RCLCPP_INFO(rclcpp::get_logger("set_control_mode_client"), "Control mode set successfully!");
    }
    else
    {
      RCLCPP_ERROR(rclcpp::get_logger("set_control_mode_client"), "Failed to set control mode!");
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("set_control_mode_client"), "Failed to call service /%s/control_selection/set_mode", namespace_.c_str());
  }

  rclcpp::shutdown();
  return 0;
}

File: python/endpoints/control_selection/set_mode.py
from clearpath_config.common.utils.yaml import read_yaml
from clearpath_control_selection_msgs.srv import SetControlMode
from clearpath_control_selection_msgs.msg import ControlMode
import rclpy
from rclpy.node import Node

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'SetControlMode'

# TODO: Enter your control mode
CONTROL_MODE: int = ControlMode.MANUAL # options: MANUAL, AUTONOMOUS, NEUTRAL. See clearpath_control_selection_msgs/msg/ControlMode.msg for more details.


class SetControlModeClient(Node):

    def __init__(self, namespace: str):
        super().__init__('set_control_mode_client')
        self._set_control_mode_client = self.create_client(SetControlMode, f'/{namespace}/control_selection/set_mode')

        while not self._set_control_mode_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().info(f'[{LOGGING_NAME}] /{namespace}/control_selection/set_mode service not available, waiting again...')

    def set_control_mode(self, mode):
        req = SetControlMode.Request()
        req.mode.mode = mode
        self.future = self._set_control_mode_client.call_async(req)
        self.future.add_done_callback(self.set_control_mode_srv_response_cb)

    def set_control_mode_srv_response_cb(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'[{LOGGING_NAME}] Control mode set successfully!')
            else:
                self.get_logger().error(f'[{LOGGING_NAME}] Failed to set control mode')
        except Exception as e:
            self.get_logger().error(f'[{LOGGING_NAME}] Service call failed: {e}')

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    set_control_mode_client = SetControlModeClient(namespace)
    set_control_mode_client.set_control_mode(CONTROL_MODE)

    rclpy.spin(set_control_mode_client)


if __name__ == '__main__':
    main()
